src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java

```diff
@@ -0,0 +1,272 @@
+                        String.format("%s / Week %s", oneDay.getSessionName(), oneWeek.getWeek());
+                labels.add(searchableLabel);
+                
+                WeeklyAdherenceReportRow row = new WeeklyAdherenceReportRow(); 
+                row.setLabel(displayLabel);
+                row.setSearchableLabel(searchableLabel);
+                row.setSessionGuid(oneDay.getSessionGuid());
+                row.setStartEventId(oneDay.getStartEventId());
+                row.setSessionName(oneDay.getSessionName());
+                row.setSessionSymbol(oneDay.getSessionSymbol());
+                row.setStudyBurstId(oneDay.getStudyBurstId());
+                row.setStudyBurstNum(oneDay.getStudyBurstNum());
+                row.setWeek(oneWeek.getWeek());
+                rows.add(row);
+            }
+        }
+
+        // Now pad the rows so they can be read straight through in any UI of the report. 
+        // Sort the rows so they display in an order that makes sense
+        List<WeeklyAdherenceReportRow> rowList = Lists.newArrayList(rows);
+        rowList.sort(ROW_COMPARATOR);
+        
+        for (int i=0; i < 7; i++) {
+            List<EventStreamDay> paddedDays = new ArrayList<>();
+            
+            for (WeeklyAdherenceReportRow row : rowList) {
+                List<EventStreamDay> days = oneWeek.getByDayEntries().get(i);
+                
+                EventStreamDay oneDay = padEventStreamDay(days, row.getSessionGuid(), row.getStartEventId());
+                LocalDate thisDate = oneWeek.getStartDate().plusDays(i);
+                oneDay.setStartDate(thisDate);
+                paddedDays.add(oneDay);
+            }
+            oneWeek.getByDayEntries().put(i, paddedDays);
+        }
+        oneWeek.setSearchableLabels(labels);
+        oneWeek.setRows(rowList);
+    }
+    
+    private EventStreamDay padEventStreamDay(List<EventStreamDay> days, String sessionGuid, String eventId) {
+        // If a session is triggered twice by two different events, it can appear twice in rows.
+        // If the events have nearly the same timestamp, the GUIDs and the labels would be exactly
+        // the same. For this reason we carry over the event ID as this identifies two unique streams
+        // and these must be tracked separately, even though they look the same in the report.
+        Optional<EventStreamDay> oneDay = days.stream()
+                .filter(day -> day.getSessionGuid().equals(sessionGuid) && day.getStartEventId().equals(eventId))
+                .findFirst();
+        if (oneDay.isPresent()) {
+            return oneDay.get();
+        }
+        return new EventStreamDay();
+    }
+    
+    private NextActivity getNextActivity(Collection<StudyReportWeek> weeks, LocalDate localToday) {
+        for (StudyReportWeek oneWeek : weeks) {
+            if (oneWeek.getStartDate().isAfter(localToday)) {
```

Comment:
Wasn't this startDate + 7 days at one point?

Reply:
This works; test added (one week's start date is after the local date, so it starts searching at the week after the current week).